File: backend/app/schemas/virtualization.py
"""虚拟化相关 Schema。"""
from __future__ import annotations

from typing import Optional, List
from datetime import datetime
from pydantic import BaseModel, Field


class EsxiHostBase(BaseModel):
    ip: str
    port: int = 443
    username: str = "root"
    password: Optional[str] = Field(
        default=None, description="可选；为空时回退到环境变量 ESXI_PASSWORD"
    )
    description: Optional[str] = None


class EsxiHostCreate(EsxiHostBase):
    probe_only: bool = False


class EsxiHostResponse(EsxiHostBase):
    id: int
    hostname: Optional[str] = None
    status: Optional[str] = None
    version: Optional[str] = None
    model: Optional[str] = None
    sort_order: Optional[int] = None
    cpu_usage: Optional[float] = None
    memory_usage: Optional[float] = None
    cpu_cores: Optional[int] = None
    memory_total_gb: Optional[float] = None
    storage_total_gb: Optional[float] = None
    storage_free_gb: Optional[float] = None
    vm_count: Optional[int] = None
    vms_running: Optional[int] = None
    last_sync_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class VirtualMachineInfo(BaseModel):
    id: Optional[str] = None
    name: str
    power_state: str = Field(description="poweredOn/poweredOff/suspended 等")
    guest_os: Optional[str] = None
    ip_address: Optional[str] = None
    description: Optional[str] = None
    instance_uuid: Optional[str] = None
    moref: Optional[str] = Field(default=None, description="Managed Object Reference (moId)")
    host_id: Optional[int] = None
    host_ip: Optional[str] = None
    cpu_count: Optional[int] = None
    memory_mb: Optional[int] = None
    cpu_usage_mhz: Optional[int] = None
    memory_usage_mb: Optional[int] = None
    uptime_seconds: Optional[int] = None
    disk_used_gb: Optional[float] = None
    disk_provisioned_gb: Optional[float] = None
    tools_status: Optional[str] = None


class VirtualMachineListResponse(BaseModel):
    total: int
    items: List[VirtualMachineInfo]


# 兼容旧设计的响应/请求，可继续复用
class VirtualMachineResponse(BaseModel):
    id: str
    uuid: str
    name: str
    host_ip: str
    status: str
    ip_address: Optional[str] = None
    os_name: Optional[str] = None
    description: Optional[str] = None
    cpu_count: int
    memory_mb: int
    tools_status: Optional[str] = None

    class Config:
        from_attributes = True


class VmListResponse(BaseModel):
    total: int
    items: List[VirtualMachineResponse]


class PowerActionRequest(BaseModel):
    action: str  # powerOn, reboot, etc.


class VMUpdateRequest(BaseModel):
    name: Optional[str] = Field(default=None, description="新名称")
    description: Optional[str] = Field(default=None, description="备注/Annotation")


class AsyncTaskResponse(BaseModel):
    task_id: str
    status: str
    message: Optional[str] = None


class VMCloneRequest(BaseModel):
    new_name: str
    target_datastore: Optional[str] = None
    power_on: bool = False
    source_ip: Optional[str] = Field(default=None, description="源虚机当前 IP，仅用于记录/日志")
    auto_config_ip: bool = Field(default=False, description="克隆后自动修改 IP（需要 Guest 凭据与 VMware Tools）")
    guest_username: Optional[str] = Field(default="root", description="Guest OS 登录账号")
    guest_password: Optional[str] = Field(default=None, description="Guest OS 登录密码")
    new_ip: Optional[str] = None
    netmask: Optional[str] = None
    gateway: Optional[str] = None
    dns: Optional[List[str]] = None
    nic_name: Optional[str] = Field(default="eth0", description="在 Guest 内的网卡名，默认 eth0；若不同需显式指定")
    disconnect_nic_first: bool = Field(default=True, description="克隆后开机前先断开网卡，避免 IP 冲突")


class VMCloneResponse(BaseModel):
    success: bool
    message: str
    new_vm_moref: Optional[str] = None
    new_vmx_path: Optional[str] = None
    ip_configured: Optional[bool] = None
    ip_message: Optional[str] = None


class VMInstallToolsRequest(BaseModel):
    ip: str = Field(..., description="SSH IP Address")
    username: Optional[str] = Field(default="root")
    password: Optional[str] = None
    credential_id: Optional[int] = None


class DatastoreStatsResponse(BaseModel):
    total_count: int
    total_capacity_gb: float
    total_free_gb: float


class HostReorderRequest(BaseModel):
    host_ids: List[int] = Field(..., description="ESXi Host ID 列表（数组顺序即最终显示顺序）")


class SuccessResponse(BaseModel):
    success: bool = True

File: frontend/src/api/virtualization.ts
import apiClient from '@/lib/axios';

// --- Interfaces ---

// 宿主机
export interface EsxiHost {
  id: number;
  ip: string;
  port?: number;
  username?: string;
  hostname: string;
  status: 'online' | 'offline' | 'auth_error';
  version?: string;
  description?: string;
  sort_order?: number;
  cpu_usage?: number; // percentage
  memory_usage?: number; // percentage
  cpu_cores?: number;
  memory_total_gb?: number;
  storage_total_gb?: number;
  storage_free_gb?: number;
  vm_count: number;
  vms_running?: number;
  last_sync?: string;
}

// Host add response（兼容探测返回 success/host_info 与正常 host 对象）
export type AddHostResponse = EsxiHost & {
  success?: boolean;
  message?: string;
  host_info?: any;
};

// 虚拟机
export interface VirtualMachine {
  id: string; // 数据库 ID
  // uuid: string; // API 可能没返回这个，或者叫 instance_uuid
  name: string;
  host_id: number;
  host_ip: string;
  power_state: string; // 'poweredOn' | 'poweredOff' | 'suspended'
  ip_address?: string;
  guest_os?: string;
  cpu_count: number;
  memory_mb: number;
  uptime_seconds?: number;
  
  // 新增运行时指标
  cpu_usage_mhz?: number;
  memory_usage_mb?: number;
  disk_used_gb?: number;
  disk_provisioned_gb?: number;
  
  description?: string; // VM 备注/Annotation
  tools_status?: string;
}

// 电源操作动作
export type VmPowerAction = 'powerOn' | 'shutdown' | 'powerOff' | 'reboot' | 'reset' | 'suspend';

// 异步任务响应
export interface AsyncTaskResponse {
  task_id: string;
  status: string;
  message?: string;
}

// Console 连接信息
export interface VmConsoleInfo {
  type: string;
  url: string;
  ticket?: string;
}

// 分页响应
export interface PageResult<T> {
  total: number;
  items: T[];
}

export interface DatastoreStats {
  total_count: number;
  total_capacity_gb: number;
  total_free_gb: number;
}

// --- API ---

export const virtualizationApi = {
  // Hosts
  getHosts: async () => {
    const response = await apiClient.get<EsxiHost[]>('/virtualization/hosts');
    return response.data;
  },

  getDatastoreStats: async () => {
    const response = await apiClient.get<DatastoreStats>('/virtualization/datastores/stats');
    return response.data;
  },

  addHost: async (data: { ip: string; port?: number; username: string; password?: string; description?: string; probe_only?: boolean }) => {
    const response = await apiClient.post<AddHostResponse>('/virtualization/hosts', data);
    return response.data;
  },

  updateHost: async (id: number, data: { ip?: string; port?: number; username?: string; password?: string; description?: string }) => {
    const response = await apiClient.put<EsxiHost>(`/virtualization/hosts/${id}`, data);
    return response.data;
  },

  deleteHost: async (id: number) => {
    await apiClient.delete(`/virtualization/hosts/${id}`);
  },

  syncHosts: async (hostId?: number) => {
    const response = await apiClient.post<{ success: boolean; message: string }>('/virtualization/sync', {
        host_id: hostId
    });
    return response.data;
  },

  reorderHosts: async (hostIds: number[]) => {
    const response = await apiClient.post<{ success: boolean; message: string }>('/virtualization/hosts/reorder', {
        host_ids: hostIds
    });
    return response.data;
  },

  // VMs
  getVms: async (params: {
    host_id?: number;
    keyword?: string;
    status?: string;
    page?: number;
    page_size?: number;
  }) => {
    const response = await apiClient.get<PageResult<VirtualMachine>>('/virtualization/vms', {
      params,
    });
    return response.data;
  },

  performPowerAction: async (vmId: string, action: VmPowerAction) => {
    const response = await apiClient.post<AsyncTaskResponse>(`/virtualization/vms/${vmId}/power`, {
      action,
    });
    return response.data;
  },

  getConsoleUrl: async (vmId: string) => {
    const response = await apiClient.get<VmConsoleInfo>(`/virtualization/vms/${vmId}/console`);
    return response.data;
  },

  updateVm: async (vmId: string, data: { name?: string; description?: string }) => {
    const response = await apiClient.patch<VirtualMachine>(`/virtualization/vms/${vmId}`, data);
    return response.data;
  },
  
  // Snapshots
  getSnapshots: async (vmId: string) => {
      // TODO: Define Snapshot interface if needed later
      const response = await apiClient.get<any[]>(`/virtualization/vms/${vmId}/snapshots`);
      return response.data;
  },

  createSnapshot: async (vmId: string, data: { name: string; description?: string }) => {
      const response = await apiClient.post<AsyncTaskResponse>(`/virtualization/vms/${vmId}/snapshots`, data);
      return response.data;
  },

  revertSnapshot: async (vmId: string, snapshotId: string) => {
      const response = await apiClient.post<AsyncTaskResponse>(`/virtualization/vms/${vmId}/snapshots/${snapshotId}/revert`);
      return response.data;
  },

  // Clone
  cloneVm: async (vmId: string, data: { 
      new_name: string; 
      target_datastore?: string; 
      power_on?: boolean;
      // IP Customization
      auto_config_ip?: boolean;
      guest_username?: string;
      guest_password?: string;
      new_ip?: string;
      netmask?: string;
      gateway?: string;
      dns?: string[]; // Array of strings
      nic_name?: string;
      disconnect_nic_first?: boolean;
      source_ip?: string; // 用于记录源虚拟机 IP
  }) => {
      const response = await apiClient.post<AsyncTaskResponse>(`/virtualization/vms/${vmId}/clone`, data);
      return response.data;
  },

  installTools: async (vmId: string, data: { ip: string; username?: string; password?: string; credential_id?: number }) => {
      const response = await apiClient.post<AsyncTaskResponse>(`/virtualization/vms/${vmId}/install-tools`, data);
      return response.data;
  }
};
